src/main/java/uk/gov/companieshouse/api/accounts/validation/StocksValidator.java

```diff
@@ -0,0 +1,306 @@
+    }
+
+    public Errors validateStocks(@Valid Stocks stocks, Transaction transaction,
+                                 String companyAccountsId,
+                                 HttpServletRequest request) throws DataException {
+
+        Errors errors = new Errors();
+
+        if (stocks.getCurrentPeriod() != null) {
+            validateCurrentPeriod(stocks, errors);
+            crossValidateCurrentPeriod(errors, request, stocks, companyAccountsId);
+        }
+
+        if (stocks.getPreviousPeriod() != null) {
+
+            try {
+                if (companyService.isMultipleYearFiler(transaction)) {
+                    validatePreviousPeriod(stocks, errors);
+                    crossValidatePreviousPeriod(errors, request, stocks, companyAccountsId);
+                } else {
+                    validateInconsistentFiling(stocks, errors);
+                }
+
+            } catch (ServiceException e) {
+                throw new DataException(e.getMessage(), e);
+            }
+        }
+        return errors;
+    }
+
+    private void validateCurrentPeriod(Stocks stocks, Errors errors) {
+        validateCurrentPeriodTotalPresent(stocks, errors);
+        validateCurrentPeriodTotalCorrect(stocks, errors);
+    }
+
+    private void validatePreviousPeriod(Stocks stocks, Errors errors) {
+        validatePreviousPeriodTotalPresent(stocks, errors);
+        validatePreviousPeriodTotalCorrect(stocks, errors);
+    }
+
+    private void validateInconsistentFiling(Stocks stocks, Errors errors) {
+
+        if (stocks.getPreviousPeriod().getPaymentsOnAccount() != null) {
+            addInconsistentDataError(errors, PREVIOUS_PERIOD_PAYMENTS_ON_ACCOUNT_PATH);
+        }
+
+        if (stocks.getPreviousPeriod().getStocks() != null) {
+            addInconsistentDataError(errors, PREVIOUS_PERIOD_STOCKS_PATH);
+        }
+
+        if (stocks.getPreviousPeriod().getTotal() != null) {
+            addInconsistentDataError(errors, PREVIOUS_PERIOD_TOTAL_PATH);
+        }
+    }
+
+    private void validateCurrentPeriodTotalPresent(Stocks stocks, Errors errors) {
+        if (stocks.getCurrentPeriod().getTotal() == null &&
+                (stocks.getCurrentPeriod().getPaymentsOnAccount() != null ||
+                        stocks.getCurrentPeriod().getStocks() != null)) {
+
+            addError(errors, invalidNote, CURRENT_PERIOD_TOTAL_PATH);
+        }
+    }
+
+    private void validatePreviousPeriodTotalPresent(Stocks stocks, Errors errors) {
+        if (stocks.getPreviousPeriod().getTotal() == null &&
+                (stocks.getPreviousPeriod().getPaymentsOnAccount() != null ||
+                        stocks.getPreviousPeriod().getStocks() != null)) {
+
+            addError(errors, invalidNote, PREVIOUS_PERIOD_TOTAL_PATH);
+        }
+    }
+
+    private void validateCurrentPeriodTotalCorrect(Stocks stocks, Errors errors) {
+
+        if (stocks.getCurrentPeriod().getTotal() == null) {
+            return;
```

Comment:
Should we be returning here? What if `total` is null but `stocks` or `paymentsOnAccount` is not null?

Reply:
This method only validates that the total value provided is correct. We call `validateCurrentPeriodTotalPresent` before this method which should take care of instances where we have data for fields other than the `total`.

In the case where we have values for `stocks` or `payments_on_account` but no `total` an `invalid_note` error would be set (in `validateCurrentPeriodTotalPresent`) for the missing total and this method (`validateCurrentPeriodTotalCorrect`) would not add any additional errors (i.e. the total is either missing or incorrect and not both).